### PR TITLE
feat : add Class Based DSU  for Path Compression O(1) & Union Rank O(logN)

### PR DESCRIPTION
**Pull request by AayushVyasKIIT**
#### Description of Change
- **DSU_path_compression.cpp** : using path compression O(1)
- **DSU_union_rank.cpp** : using union rank to keep track of parent[i] in O(logN).
- Class based to provide accessibility to operations efficiently.
- Also housekeeping for **maximal** and **minimal** elements of each set(chunk).
- Documentation and explanation of the concept as code progresses with test case and implementation.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
This a pretty detailed implementation of Disjoint set unions, which is very useful in competitive programming . Using class based structure, this file can be directly imported and is ready to be used in other programs.
PS: if you want to import this, the main function has to be removed first, I have included it because of the guidelines for PR.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1571"></a>

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Still not up to the [repository standards](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md). You can check the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) for a prototype of how to make your file. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please check the [typical structure of a program](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program).

File: data_structures/dsu_path_compression.cpp [hunk 1]
```diff
@@ -1,22 +1,46 @@
-#include <iostream>
-#include <vector>
+/**
+ * @file
+ * @brief [DSU(Disjoint sets)](https://en.wikipedia.org/wiki/Disjoint-set-data_structure)
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:3]:
```suggestion
 * @brief [DSU (Disjoint sets)](https://en.wikipedia.org/wiki/Disjoint-set-data_structure)
```

File: data_structures/dsu_path_compression.cpp [hunk 2]
```diff
@@ -1,22 +1,46 @@
...
+ * @details
+ * dsu : It is a very powerful data structure which keeps track of different
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:5]:
```suggestion
 * DSU: It is a very powerful data structure that keeps track of different
```

File: data_structures/dsu_path_compression.cpp [hunk 3]
```diff
@@ -1,22 +1,46 @@
...
+ * 3) get_max(i),get_min(i) : returns the maximum and minimum 
+ * Below is the class-based approach which uses the heuristic of path compression.
+ * Using path compression in findSet(i),we are able to get to the representative of i
+ * in O(1) time.
+ * @author [AayushVyasKIIT](https://github.com/AayushVyasKIIT)
+ * @see dsu_union_rank.cpp
+ */
+
+#include <iostream> ///for io
+#include <vector>   ///for using vectors
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:21]:
```suggestion
#include <iostream> /// for IO operations
#include <vector>   /// for std::vector
```

File: data_structures/dsu_path_compression.cpp [hunk 4]
```diff
@@ -1,22 +1,46 @@
...
+ */
+class dsu{
     private:
-        // p: keeps track of parent of i
-        // depth: tracks the depth of i
-        // setSize: size of each chunk(set)
-        // maxElement : max of each set, using maxElement[representative]
-        // minElement : min of each set, using minElement[representative]
-        vector<int> p,depth,setSize,maxElement,minElement;
+        vector<int> p; ///<keeps track of the parent of ith element
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:33]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)).

File: data_structures/dsu_path_compression.cpp [hunk 5]
```diff
@@ -75,37 +109,61 @@ class DSU{
...
             minElement[y] = std::min(minElement[x],minElement[y]);
         }
-        //checks if both belongs to same set
+        /**
+         * @brief A utility function which check whether i and j belongs to
+         * same set or not
+         * @param i element of some set
+         * @param j element of some set
+         * @returns `true` if element i and j are in same set
+         * @returns `false` if element i and j are not in same set
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:118]:
```suggestion
         * @returns `true` if element `i` and `j` ARE in the same set
         * @returns `false` if element `i` and `j` are NOT in same set
```

File: data_structures/dsu_union_rank.cpp [hunk 6]
```diff
@@ -1,20 +1,45 @@
-#include <iostream>
-#include <vector>
+/**
+ * @file
+ * @brief [dsu(Disjoint sets)](https://en.wikipedia.org/wiki/Disjoint-set-data_structure)
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:3]:
```suggestion
 * @brief [DSU (Disjoint sets)](https://en.wikipedia.org/wiki/Disjoint-set-data_structure)
```

File: data_structures/dsu_union_rank.cpp [hunk 7]
```diff
@@ -1,20 +1,45 @@
...
+ * 3) getParents(i): prints the parent of i and so on and so forth.
+ * Below is the class-based approach which uses the heuristic of union-ranks.
+ * Using union-rank in findSet(i),we are able to get to the representative of i
+ * in slightly delayed O(logN) time but it allows us to keep tracks of the parent of i.
+ * @author [AayushVyasKIIT](https://github.com/AayushVyasKIIT)
+ * @see dsu_path_compression.cpp
+ */
+
+#include <iostream> ///for io
+#include <vector> ///for using vectors
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:21]:
```suggestion
#include <iostream> /// for IO operations
#include <vector>  /// for std::vector
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Good work! 🙂👍

File: data_structures/dsu_path_compression.cpp [hunk 8]
```diff
@@ -0,0 +1,183 @@
+/**
+ * @file
+ * @brief [DSU (Disjoint sets)](https://en.wikipedia.org/wiki/Disjoint-set-data_structure)
+ * @details
+ * DSU: It is a very powerful data structure that keeps track of different
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:5]:
```suggestion
 * It is a very powerful data structure that keeps track of different
```

File: data_structures/dsu_path_compression.cpp [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        /**
+         * @brief A utility function that returns the min element of the set to which i belongs to
+         * @param i element of some set
+         * @returns minimum of the set to which i belongs to 
+         */
+        int get_min(int i){
+            return minElement[findSet(i)];
+        }
+
+};
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:159]: Add self-test implementations using `assert` (see the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program as an example).

File: data_structures/dsu_union_rank.cpp [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        void getParents(int i){
+            while(p[i]!=i){
+                cout << i << " ->";
+                i = p[i];
+            }
+            cout << p[i] << endl;
+        }
+        
+};
+
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:123]: Add self-test implementations using `assert` (see the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program as an example).

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! Almost there! 😄

File: data_structures/dsu_path_compression.cpp [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+ * @brief Disjoint sets union data structure, class based representation.
+ * @param n number of elements
+ */
+class dsu{
+    private:
+        vector<int> p; ///<keeps track of the parent of ith element
+        vector<int> depth; ///<tracks the depth(rank) of i in the tree
+        vector<int> setSize;///<size of each chunk(set)
+        vector<int> maxElement;/// <maximum of each set to which i belongs to
+        vector<int> minElement;/// <minimum of each set to which i belongs to
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:37]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). Let me know if you need any help. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: data_structures/dsu_path_compression.cpp (hunk 11 above)
Comment (AayushVyasKIIT) [reply data_structures/dsu_path_compression.cpp]: Did it!

File: data_structures/dsu_union_rank.cpp (hunk 10 above)
Comment (AayushVyasKIIT) [reply data_structures/dsu_union_rank.cpp]: Done!

File: data_structures/dsu_path_compression.cpp [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+ * Below is the class-based approach which uses the heuristic of path compression.
+ * Using path compression in findSet(i),we are able to get to the representative of i
+ * in O(1) time.
+ * @author [AayushVyasKIIT](https://github.com/AayushVyasKIIT)
+ * @see dsu_union_rank.cpp
+ */
+
+#include <iostream> /// for IO operations
+#include <vector>   /// for std::vector
+#include <cassert>   /// for assert
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:22]:
```suggestion
#include <iostream>  /// for IO operations
#include <vector>    /// for std::vector
#include <cassert>  /// for assert
```

File: data_structures/dsu_union_rank.cpp [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+ * Below is the class-based approach which uses the heuristic of union-ranks.
+ * Using union-rank in findSet(i),we are able to get to the representative of i
+ * in slightly delayed O(logN) time but it allows us to keep tracks of the parent of i.
+ * @author [AayushVyasKIIT](https://github.com/AayushVyasKIIT)
+ * @see dsu_path_compression.cpp
+ */
+
+#include <iostream> /// for IO operations
+#include <vector>  /// for std::vector
+#include <cassert>   /// for assert
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:22]:
```suggestion
#include <iostream>   /// for IO operations
#include <vector>    /// for std::vector
#include <cassert>  /// for assert
```

File: data_structures/dsu_path_compression.cpp [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+        vector<uint64_t> maxElement;/// <maximum of each set to which i belongs to
+        vector<uint64_t> minElement;/// <minimum of each set to which i belongs to
+    public:
+        /**
+        * @brief contructor for initialising all data members.
+        * @param n number of elements
+        */
+        explicit dsu(uint64_t n){
+            p.assign(n,0);
+            //initially all of them are their own parents.
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:46]:
```suggestion
            /// initially, all of them are their own parents
```

File: data_structures/dsu_path_compression.cpp [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+            for(uint64_t i=0;i<n;i++){
+                p[i] = i;
+            }
+            //initially all have depth =0
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:50]:
```suggestion
            /// initially all have depth are equals to zero
```

File: data_structures/dsu_path_compression.cpp [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+            depth.assign(n,0);
+            maxElement.assign(n,0);
+            minElement.assign(n,0);
+            for(uint64_t i=0;i<n;i++){
+                depth[i] = 0;
+                maxElement[i] = i;
+                minElement[i] = i;
+            }
+            setSize.assign(n,0);
+            //initially set size will be 1
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:60]:
```suggestion
            /// initially set size will be equals to one
```

File: data_structures/dsu_path_compression.cpp [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+        }
+        /**
+         * @brief Method that combines two disjoint sets to which i and j belongs to 
+         * and make a single set having a common representative.
+         * @param i element of some set
+         * @param j element of some set
+         * @returns void
+        */
+        void UnionSet(uint64_t i,uint64_t j){
+            //check if both belongs to same set or not
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:86]:
```suggestion
            /// check if both belongs to the same set or not
```

File: data_structures/dsu_path_compression.cpp [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+            if(isSame(i,j)){
+                return;
+            }
+
+            //we find the representative of the i and j
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:91]:
```suggestion
            /// we find the representative of `i` and `j`
```

File: data_structures/dsu_path_compression.cpp [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+            uint64_t x = findSet(i);
+            uint64_t y = findSet(j);
+
+            //always keeping the min as x
+            //shallow tree
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:96]:
```suggestion
            /// always keeping the min as x
            /// shallow tree
```

File: data_structures/dsu_path_compression.cpp [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+            if(depth[x]>depth[y]){
+                std::swap(x,y);
+            }
+            //making the shallower root's parent the deeper root
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:100]:
```suggestion
            /// making the shallower root's parent the deeper root
```

File: data_structures/dsu_path_compression.cpp [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+            p[x] = y;
+            
+            //if same depth then increase one's depth
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:103]:
```suggestion
            /// if same depth, then increase one's depth
```

File: data_structures/dsu_path_compression.cpp [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+            if(depth[x] == depth[y]){
+                depth[y]++;
+            }
+            //total size of the resultant set.
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:107]:
```suggestion
            /// total size of the resultant set
```

File: data_structures/dsu_path_compression.cpp [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+            //making the shallower root's parent the deeper root
+            p[x] = y;
+            
+            //if same depth then increase one's depth
+            if(depth[x] == depth[y]){
+                depth[y]++;
+            }
+            //total size of the resultant set.  
+            setSize[y] += setSize[x];
+            //changing the maximum elements
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:109]:
```suggestion
            /// changing the maximum elements
```

File: data_structures/dsu_path_compression.cpp [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+    d.UnionSet(1,2); //performs union operation on 1 and 2
+    d.UnionSet(1,4); //performs union operation on 1 and 4
+    vector<uint64_t> ans = {1,4,3};
+    for(uint64_t i=0;i<ans.size();i++){
+        assert(d.get(4).at(i) == ans[i]); //makes sure algorithm works fine
+    }
+    cout << "Test case# 1: passed"<<endl;
+}
+/**
+ * @brief Self-implementation Test case #2
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:184]:
```suggestion
 * @brief Self-test implementations, 2nd test
```

File: data_structures/dsu_path_compression.cpp [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+         * @param i element of some set
+         * @returns minimum of the set to which i belongs to 
+         */
+        uint64_t get_min(uint64_t i){
+            return minElement[findSet(i)];
+        }
+
+};
+/**
+ * @brief Self-implementation Test case #1
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:167]:
```suggestion
 * @brief Self-test implementations, 1st test
```

File: data_structures/dsu_path_compression.cpp [hunk 26]
```diff
@@ -0,0 +1,215 @@
...
+ * */
+int main(){
+    uint64_t n = 10;///< number of items
+    dsu d(n+1);///< object of class disjoint sets
+
+    test1(); //< test case# 1
+    test2(); //< test case# 2
+    
+    return 0;
+}
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:215]:
```suggestion
}

```

File: data_structures/dsu_path_compression.cpp [hunk 27]
```diff
@@ -0,0 +1,215 @@
...
+    d.UnionSet(3,5);
+    d.UnionSet(5,6);
+    d.UnionSet(5,7);
+    vector<uint64_t> ans = {3,7,4};
+    for(uint64_t i=0;i<ans.size();i++){
+        assert(d.get(3).at(i) == ans[i]); //makes sure algorithm works fine
+    }
+    cout << "Test case# 2: passed"<<endl;
+}
+
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:201]:
```suggestion
```

File: data_structures/dsu_path_compression.cpp [hunk 28]
```diff
@@ -0,0 +1,215 @@
...
+    dsu d(n+1);///< object of class disjoint sets
+    //set 1
+    d.UnionSet(3,5);
+    d.UnionSet(5,6);
+    d.UnionSet(5,7);
+    vector<uint64_t> ans = {3,7,4};
+    for(uint64_t i=0;i<ans.size();i++){
+        assert(d.get(3).at(i) == ans[i]); //makes sure algorithm works fine
+    }
+    cout << "Test case# 2: passed"<<endl;
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:199]:
```suggestion
    cout << "2nd test passed!"<<endl;
```

File: data_structures/dsu_path_compression.cpp [hunk 29]
```diff
@@ -0,0 +1,215 @@
...
+    uint64_t n = 10;///< number of items
+    dsu d(n+1);///< object of class disjoint sets
+    //set 1
+    d.UnionSet(1,2); //performs union operation on 1 and 2
+    d.UnionSet(1,4); //performs union operation on 1 and 4
+    vector<uint64_t> ans = {1,4,3};
+    for(uint64_t i=0;i<ans.size();i++){
+        assert(d.get(4).at(i) == ans[i]); //makes sure algorithm works fine
+    }
+    cout << "Test case# 1: passed"<<endl;
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:181]:
```suggestion
    cout << "1st test passed!"<<endl;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/AayushVyasKIIT/C-Plus-Plus/actions

**Comment by AayushVyasKIIT** [comment]
- [x] Did all the changes requested. Which are marked as outdated right now.

 >  outdated
 
- [x] Enabled github actions.

File: data_structures/dsu_path_compression.cpp [hunk 30]
```diff
@@ -0,0 +1,212 @@
...
+ * @brief Disjoint sets union data structure, class based representation.
+ * @param n number of elements
+ */
+class dsu {
+ private:
+    vector<uint64_t> p;           ///< keeps track of the parent of ith element
+    vector<uint64_t> depth;       ///< tracks the depth(rank) of i in the tree
+    vector<uint64_t> setSize;     ///< size of each chunk(set)
+    vector<uint64_t> maxElement;  /// <maximum of each set to which i belongs to
+    vector<uint64_t> minElement;  /// <minimum of each set to which i belongs to
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:39]:
```suggestion
    vector<uint64_t> maxElement;  ///< maximum of each set to which i belongs to
    vector<uint64_t> minElement;  ///< minimum of each set to which i belongs to
```

File: data_structures/dsu_path_compression.cpp [hunk 31]
```diff
@@ -0,0 +1,212 @@
...
+     * @returns minimum of the set to which i belongs to
+     */
+    uint64_t get_min(uint64_t i) { return minElement[findSet(i)]; }
+};
+/**
+ * @brief Self-test implementations, 1st test
+ * @returns void
+ */
+static void test1() {
+    /* the minimum, maximum and size of the set*/
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:169]:
```suggestion
    // the minimum, maximum, and size of the set
```

File: data_structures/dsu_path_compression.cpp [hunk 32]
```diff
@@ -0,0 +1,212 @@
...
+        assert(d.get(4).at(i) == ans[i]);  // makes sure algorithm works fine
+    }
+    cout << "1st test passed!" << endl;
+}
+/**
+ * @brief Self-implementations, 2nd test
+ * @returns void
+ */
+static void test2() {
+    /* the minimum, maximum and size of the set */
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:186]:
```suggestion
    // the minimum, maximum, and size of the set 
```

File: data_structures/dsu_union_rank.cpp [hunk 33]
```diff
@@ -0,0 +1,183 @@
...
+
+/**
+ * @brief Disjoint sets union data structure, class based representation.
+ * @param n number of elements
+ */
+class dsu{
+    private:
+        vector<uint64_t> p; ///<keeps track of the parent of ith element
+        vector<uint64_t> depth; ///<tracks the depth(rank) of i in the tree
+        vector<uint64_t> setSize;///<size of each chunk(set)
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:37]:
```suggestion
        vector<uint64_t> p;          ///< keeps track of the parent of ith element
        vector<uint64_t> depth;      ///< tracks the depth(rank) of i in the tree
        vector<uint64_t> setSize;    ///< size of each chunk(set)
```

File: data_structures/dsu_union_rank.cpp [hunk 34]
```diff
@@ -0,0 +1,183 @@
...
+    dsu d(n+1); ///< object of class disjoint sets
+    d.unionSet(2,1); ///< performs union operation on 1 and 2
+    d.unionSet(1,4);
+    d.unionSet(8,1);
+    d.unionSet(3,5);
+    d.unionSet(5,6);
+    d.unionSet(5,7);
+    d.unionSet(9,10);
+    d.unionSet(2,10);
+    ///keeping track of the changes using parent pointers
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:142]:
```suggestion
    // keeping track of the changes using parent pointers
```

File: data_structures/dsu_union_rank.cpp [hunk 35]
```diff
@@ -0,0 +1,183 @@
...
+    vector<uint64_t> ans = {7,5};
+    for(uint64_t i=0;i<ans.size();i++){
+        assert(d.getParents(7).at(i) == ans[i]); ///makes sure algorithm works fine
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:145]:
```suggestion
        assert(d.getParents(7).at(i) == ans[i]); // makes sure algorithm works fine
```

File: data_structures/dsu_union_rank.cpp [hunk 36]
```diff
@@ -0,0 +1,183 @@
...
+    }
+    cout << "1st test passed!"<<endl;
+}
+/**
+ * @brief Self-implementations, 2nd test
+ * @returns void
+ */
+static void test2() {
+    /* checks the parents in the resultant structures */
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:154]:
```suggestion
    // checks the parents in the resultant structures 
```

File: data_structures/dsu_union_rank.cpp [hunk 37]
```diff
@@ -0,0 +1,183 @@
...
+    }
+    cout << "2nd test passed!"<<endl;
+}
+/**
+ * @brief Main function
+ * @returns 0 on exit 
+ */
+int main(){
+    test1();
+    test2();
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:179]:
```suggestion
    test1();  // run 1st test
    test2();  // run 2nd test
```

File: data_structures/dsu_union_rank.cpp [hunk 38]
```diff
@@ -0,0 +1,183 @@
...
+
+
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:181]:
```suggestion
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: data_structures/dsu_path_compression.cpp [hunk 39]
```diff
@@ -0,0 +1,212 @@
...
+     * @returns minimum of the set to which i belongs to
+     */
+    uint64_t get_min(uint64_t i) { return minElement[findSet(i)]; }
+};
+/**
+ * @brief Self-test implementations, 1st test
+ * @returns void
+ */
+static void test1() {
+    /* the minimum, maximum and size of the set*/
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:169]:
```suggestion
    // the minimum, maximum, and size of the set
```

File: data_structures/dsu_union_rank.cpp (hunk 33 above)
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:37]:
```suggestion
        vector<uint64_t> p;          ///< keeps track of the parent of ith element
        vector<uint64_t> depth;      ///< tracks the depth(rank) of i in the tree
        vector<uint64_t> setSize;    ///< size of each chunk(set)
```

File: data_structures/dsu_union_rank.cpp (hunk 34 above)
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:142]:
```suggestion
    // keeping track of the changes using parent pointers
```

File: data_structures/dsu_union_rank.cpp [hunk 40]
```diff
@@ -0,0 +1,183 @@
...
+        assert(d.getParents(7).at(i) == ans[i]); // makes sure algorithm works fine
+    }
+    cout << "1st test passed!"<<endl;
+}
+/**
+ * @brief Self-implementations, 2nd test
+ * @returns void
+ */
+static void test2() {
+    /* checks the parents in the resultant structures */
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:154]:
```suggestion
    // checks the parents in the resultant structures 
```

File: data_structures/dsu_union_rank.cpp [hunk 41]
```diff
@@ -0,0 +1,183 @@
...
+    }
+    cout << "2nd test passed!"<<endl;
+}
+/**
+ * @brief Main function
+ * @returns 0 on exit 
+ */
+int main(){
+    test1();
+    test2();
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:179]:
```suggestion
    test1();  // run 1st test
    test2();  // run 2nd test
```

File: data_structures/dsu_union_rank.cpp [hunk 42]
```diff
@@ -0,0 +1,183 @@
...
+
+
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:181]:
```suggestion
```

File: data_structures/dsu_union_rank.cpp (hunk 33 above)
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:37]:
```suggestion
        vector<uint64_t> p;          ///< keeps track of the parent of ith element
        vector<uint64_t> depth;      ///< tracks the depth(rank) of i in the tree
        vector<uint64_t> setSize;    ///< size of each chunk(set)
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: data_structures/dsu_path_compression.cpp [hunk 43]
```diff
@@ -0,0 +1,212 @@
...
+     */
+    uint64_t get_max(uint64_t i) { return maxElement[findSet(i)]; }
+    /**
+     * @brief A utility function that returns the min element of the set to
+     * which i belongs to
+     * @param i element of some set
+     * @returns minimum of the set to which i belongs to
+     */
+    uint64_t get_min(uint64_t i) { return minElement[findSet(i)]; }
+};
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:163]:
```suggestion
};

```

File: data_structures/dsu_path_compression.cpp [hunk 44]
```diff
@@ -0,0 +1,212 @@
...
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ * */
+int main() {
+    uint64_t n = 10;  ///< number of items
+    dsu d(n + 1);     ///< object of class disjoint sets
+
+    test1();  //< test case# 1
+    test2();  //< test case# 2
```
Comment (Panquesito7) [inline data_structures/dsu_path_compression.cpp:209]:
```suggestion
    test1();  // run 1st test case
    test2();  // run 2nd test case
```

File: data_structures/dsu_union_rank.cpp [hunk 45]
```diff
@@ -0,0 +1,187 @@
...
+               ans[i]);  // makes sure algorithm works fine
+    }
+    cout << "1st test passed!" << endl;
+}
+/**
+ * @brief Self-implementations, 2nd test
+ * @returns void
+ */
+static void test2() {
+    /* checks the parents in the resultant structures */
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:158]:
```suggestion
    // checks the parents in the resultant structures 
```

File: data_structures/dsu_union_rank.cpp [hunk 46]
```diff
@@ -0,0 +1,187 @@
...
+    }
+    cout << "2nd test passed!" << endl;
+}
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test1();
+    test2();
```
Comment (Panquesito7) [inline data_structures/dsu_union_rank.cpp:184]:
```suggestion
    test1();  // run 1st test
    test2();  // run 2nd test
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Pull request merged**